Refactor lenient in strict out

erikkemperman opened the pull request:
Cf. [#4](https://github.com/gulpjs/value-or-function/issues/4).

I thought I'd submit what I have at this point, probably not going to be able to work on this further the next couple of days.

When looking at a project with fresh eyes I tend to have too many ideas, probably some that occurred to you but were rejected. Still, figured I'd list some of them here:
- Is returning `null` really the most convenient way to signal rejected values? Maybe allow the caller to provide a fallback/default, or even throw a `TypeError`?
- I played around with passing the given value to the uppercase functions like `Number` et. al., without the `new` keyword, to do coercion. However that would probably be _too_ lenient, e.g. `Boolean('foo')` would be accepted. Still, I'd like it if string arguments were supported, for instance if I want to provide options read from a .json file. Unfortunately I don't see clean way to do this without having to implement some functions separately for the various types. Would that be too ugly?
- Would it make sense to support `integer` and `float` separately (where the latter is just an alias for `number`)?
- Out of curiosity, what possible use do you see for type `undefined`?

erikkemperman commented:
Sorry, forgot to mention: I realize that some changes to the README would be in order if this, or something like it, were merged, but thought it would be better to wait for some feedback on the above.

erikkemperman commented:
@phated I added a bunch of commits to this PR, I'd consider all of these part of the "lenient in, strict out" issue (#4).

I am still thinking it might make sense to
- split float and integer coercion
- split timestamp and date coercion
- allow the user to specify fallback value (to be returned instead of `null`)

The first two of these would only expand and not break existing callers. The last one would.

Thoughts?

(Edit: adapted README as well.)

phated commented:
Totally missed that this PR came in.  I don't have time to review today but I'll try within the next few days.  Sorry about that.

erikkemperman commented:
No worries, please! I happen to have a little time at the moment to play around with these things, but I'm aware I've been taking up some of yours in the process.

erikkemperman commented:
FYI I polished some of these commits, will leave this branch alone now.

phated commented on index.js:
```diff
@@ -6,57 +6,125 @@ var types = [
 
-  if (!isType && !isFunction) {
+function primitive(value) {
+  if (value !== null && typeof value === 'object') {
```
should this be `!=` to handle a value of `undefined`?

erikkemperman replied:
That would result in an error for `undefined` values? The purpose of this function is just to call `.valueOf` for things like `new String()` and `new Number()`, and I think that it returning `undefined` if the value is `undefined` is reasonable, it leaves other non-objects alone as well.

phated replied:
The `typeof ... 'object'` should filter out `undefined`

phated replied:
This would need to check that `valueOf` exists and is a function also.  `Object.create(null)` results in an object without `valueOf`

erikkemperman replied:
Ah! overlooked that possibility. I will also need a guard against `.toString` being absent.

I'll have some time to fix things up tomorrow.

erikkemperman replied:
True that `typeof undefined` would not trigger the `valueOf`, but changing `!==` to `!=` doesn't alter the behaviour for the `undefined` (or any other) case... So may as well do the cheaper triple equals. Am I missing something? (It's been a long day :-))

phated replied:
Often, `!=` is faster than triple equal (haven't test this in a few years).  You are correct that `!=` and `!==` would act the same in the current statement.  If `!=` is used, the statement wouldn't have to call `typeof` on undefined.

phated replied:
Sounds good on fixing things up.  I'm out the rest of the week so I'll take a look sometime next week.

phated commented:
@erikkemperman the `coerce.*` methods seem overly complex.  Remember: these are called on each iteration of a stream in gulp so we want to process as quickly as possible (even at the cost of being "less correct").

erikkemperman commented:
Thanks for taking the time. It can probably be simplified a little, I'll take a look!

For what it's worth, though, I tried to make a point of having the `coerce.*` functions handle their expected case first, so per call to `normalize.*` I don't think there's too many more statements in common scenarios...

Also, while it is not exactly obvious from how github renders these commits in aggregate, the `normalize` function itself became somewhat simpler (and ceases to invoke `coerce` twice for function values): https://github.com/gulpjs/value-or-function/pull/5/commits/2dbfed47ab4c8b57bd79914b35987cbea9f665ba

The admittedly buttugly bits are coercing strings to some other type. I was not happy with the native coercers; e.g. `Boolean(<string>)` would accept `'maybe'` (as true), and `Number(<string>)` would reject `'infinity'` (it wants uppercase).

phated commented:
I think calling `.valueOf` on objects created by constructors (that result in PrimitiveObjects in chrome) should be enough and we don't need to accept things like strings of `true` or `t`, etc.

erikkemperman commented:
Fair enough, that will get rid of the ulgy bits mostly. Callers will just have to do their own conversion if they have strings, then.

Should `.date` accept strings?

phated commented:
I don't think `.date()` should accept strings, only numbers/number objects/date objects.  If you have a better name (we couldn't think of one) for `.date()`, I'd be fine with renaming it.

erikkemperman commented:
@phated So I've simplified, and squashed everything into a single commit. I've updated the PR to reflect what I ended up with, but I'm not entirely sure about this...

For one thing, it is now somewhat less forgiving of input types as I had originally imagined it might be, by not accepting strings for the built-in types (I was thinking that command-line options and configuration files might be a common source of strings that folks would like to coerce).

Of course it's a perfectly reasonable argument that that kind of thing should be handled externally if needed. On the other hand, perhaps I was just being much too particular about the string values I wanted to accept, and a simpler approach (just calling the native coercers `Number(...)` et. al.) might have been more palatable?

Compared to master, this branch effectively just adds support for `new Number`, `new Boolean` and `new String`. And I guess it cleans up `.date` a little in that it is now integrated with the other types and thus satisfies `vof.normalize('date', ...) === vof.normalize.date(...)`. If this smaller change had been my target from the start, perhaps I would have done it differently, or even decided the benefit doesn't outweigh the burden of added complexity.

So I'm not sure if/how to proceed with this, curious if you have ideas.

Sorry for being so verbose again, but two more general thoughts I had going over this, which apply to current master as well I suppose:

Does it actually make sense to expose `normalize.function`? It complicates handling of function-values (more so with `coerce` logic than it did with `conforms`, but still).

I would think that `null` ought to be valid input for `normalize.object`, but that's a little bit awkward in that we can't simply conclude that a `null` return value signals unacceptable input. Maybe I am overlooking some intended purpose for `normalize.undefined` but if not, I might propose to remove it and adopt `undefined` as unambiguous return value for bad input?

PS I just noticed that coverage went down, I'll take a quick look at that now.

erikkemperman commented:
Minor tweak to the commit, coverage back up.

phated commented on index.js:
```diff
@@ -1,62 +1,109 @@
+function normalize(coercer, value) {
+  if (typeof value === 'function') {
+    if (coercer === 'function') {
+      return value;
```
why is the value returned if both are a function?

phated replied:
Nevermind, didn't realize it was checking if it is a string of `'function'`

phated commented on index.js:
```diff
@@ -1,62 +1,109 @@
+
+coerce.string = function(value) {
+  value = primitive(value);
+  if (value != null &&
```
What case does this handle that `primitive` doesn't?  Maybe I'll find out in the tests.

erikkemperman replied:
For `Object`s -- other than `String` -- this calls `.toString()`, just like regular coercion through e.g. concatenation of non-string objects. Maybe it's too much, seemed like a good idea at the time.

phated replied:
But won't the `primitive` call above it turn the object into whatever `valueOf` returns?  Wouldn't that have to return another object with the `toString` method?

erikkemperman replied:
Ah, right. I originally had the `primitive` call after checking if there's a `toString` method. That did not have the problem you point out here, but in amending the commit this got turned around somehow. The default implementation for `.valueOf` just returns the object itself, so the object used in the accompanying test case did not throw up this possibility. I will sort this out.

phated commented on index.js:
```diff
@@ -1,62 +1,109 @@
-function conforms(type, value) {
-  if (typeof type === 'string') {
-    return typeof value === type;
+function fallback(type, value) {
```
why is this named `fallback`?  Doesn't seem to describe what it does.

erikkemperman replied:
It's what is used when there's no custom or built-in coercer for the given type. Not a great name, I agree. At first I didn't have a separate function for it and thus no need to make up a name, but it got repeated a couple of times and I figured I'd better separate it. Suggestions for a better name?

phated replied:
maybe just name it `typeOf` since that's what it does and how it is used.  `typeOf('string', value)`

erikkemperman replied:
Well, it doesn't return the type, which is what I would expect of a function called `typeOf`, but it returns the value if the type matches. Still it's better than `fallback` I suppose, so I'll use this ftm.

phated commented on test/index.js:
```diff
@@ -56,76 +56,78 @@ describe('normalize', function() {
-      return (typeof value === 'string');
-    };
-    var value = function() {
+  it('returns null if return value does not satisfy coercer', function(done) {
```
it doesn't technically return null now, you have to force it to be null.  What happens if `null` in the coercer were `undefined` or `'fail'`

erikkemperman replied:
I'm not entirely sure what you mean... The "contract" for custom coercers would be to return a value or `null` to indicate rejection. Are you worried about coercers that do otherwise?

phated replied:
I understand that is the implied contract, but the test name seems to imply that `null` is always returned on the non-satisfy.  Maybe falsey return values should be transformed into the actual `null` value?  Or maybe the test title just needs to be clarified.

erikkemperman replied:
My understanding is this particular test just verifies that the return value of the (function-)value is passed through a custom coercer, if any. I'll try to clarify the title.

I don't think it should transform falsey return values from custom coercers to `null` -- what if they want to accept `false` or `''` as valid?

phated commented on test/index.js:
```diff
@@ -240,14 +320,22 @@ describe('normalize.undefined', function() {
     done();
   });
+
+  it('rejects values that won\'t coerce to undefined', function(done) {
```
I think this is supposed to say "function" instead of "undefined"

erikkemperman replied:
I don't think so. GH obscured the separation here, but this test is in the `normalize.undefined` section.

phated replied:
Ah,  I missed the collapsed button.  I assumed it was supposed to be added to the `function` tests because you didn't add a "rejects" test to that block like you did for every other one.

erikkemperman replied:
I'll add a reject test for `function` as well.

phated commented on test/index.js:
```diff
@@ -240,14 +320,22 @@ describe('normalize.undefined', function() {
+
+  it('rejects values that won\'t coerce to undefined', function(done) {
+    var value = 'invalid';
+    var result = normalize.boolean(value);
```
I think this should be `normalize.function(value)`.  Probably copy-paste oversight.

erikkemperman replied:
Ugh that is a sloppy/paste, indeed. It should have been `normalize.undefined(value)` though, not `.function` (see comment above). I'll fix it and force-push.

phated commented on test/index.js:
```diff
@@ -240,14 +320,22 @@ describe('normalize.undefined', function() {
 
 describe('normalize.date', function() {
 
   it('compares value to typeof number', function(done) {
```
this probably needs rewording

erikkemperman replied:
You mean the test descriptions with `compares to typeof`? I thought the same, but then I wasn't sure what to change it to that would keep clear the distinction between, e.g., `typeof number` and `instanceof Number`. I thought only the uppercase would be too minute. Do you have a suggestion?

phated replied:
"coerces a number to a Date object" maybe?

phated commented:
@erikkemperman I like where this has ended up.  Commented inline.

erikkemperman commented:
@phated Thanks for the feedback! I'm still wondering if it makes sense to have `.function` and `.undefined`, and if the latter would go away I might argue that `undefined` would be a slightly more convenient return value to signal rejection (`null` is a valid value for type `'object'`). Thoughts?

phated commented:
I think `.function` still makes sense.  I'm not sure about `.undefined` but signals a different behavior when used in something like `_.assign()` so maybe the distinction is still needed (not sure).

erikkemperman commented:
Amended according to comments. 

I guess I just lack the fantasy to see the use case for `.function` :-) If this were called with a non-function, would you expect it to coerce? E.g. that `.function('foo')` would be coerced to a function returning `'foo'`? 

If there's a non-academic case for it, fair enough, but otherwise some simplifications could be done.

I'm not sure what your point is about lodash semantics, it just seemed to me that it might be convenient to specify a return value for rejections that isn't sometimes an accepted value (like `null` from `.object()`).

phated commented:
I don't think `.function('foo')` should be coerced.  I think it should reject since it isn't a function.  In the current gulp use cases, we call each function normalize inside the stream and it can be called with every file but I imagine a use case where the normalization of options happens when first passed in and then used in each iteration:

``` js
var func = require('value-or-function').function;
function filterStream(opts) {
  var filter = func(opts.filter);

  return through.obj(function (file, enc, cb) {
    if (filter(file)) {
      cb(null, file);
    } else {
      cb();
    }
  });
}

var opts = {
  filter: function(file) {
    if (file.isStream()) {
      return false;
    }
    return true;
  }
};
var s = filterStream(opts);
```

phated commented:
I don't think that `null` should satisfy `object`.  If someone wants that, they need to use a custom coercer.

erikkemperman commented:
I would expect it to accept `null` from `.object()`, actually, but maybe that's just me. Currently it is accepted. But if it were rejected then I think it's all good, `null` would then unambiguously signal rejection. All right, I'll amend the commit to make it so!

I was playing around with options handling in vinyl-fs earlier, and have some work in progress to merge what I was doing with the ongoing effort to split off `vinyl-prepare`. I think the direction I was headed might mesh nicely with the pattern you sketch above.

More or less in parallel, I have also looked into doing the same split for `vfs.src`. If you're interested I'll make some time to clean that up a little and share.

phated commented:
@erikkemperman awesome!  `vinyl-prepare` actually needs separate methods for `src` and `symlink`.  I was working on `symlink` this week in relation to some vinyl changes that need to happen.  Can't wait to see your results.

erikkemperman commented:
Amended the commit (of course I only had to change the descriptions of the tests, the return value was already `null` :-))

phated commented on index.js:
```diff
@@ -1,62 +1,108 @@
 
-  if (conforms(type, result)) {
-    return result;
+    return value.toString();
```
`.toString()` doesn't necessarily have to return a string. You might want to do `value = value.toString()` here instead of returning it.

phated commented:
@erikkemperman just one last thing (and a test please) then I think this is complete.

erikkemperman committed:
Lenient input, strict output

erikkemperman commented:
I actually had `value = value.toString()` before shuffling the `primitive` call to after invoking `toString` :-)

Thought it didn't make sense to coerce the result of `toString` so I made it return immediately. But you're right, it shouldn't count on it always returning a string.

Amended. Now I really have to get some sleep, but cool if I got the last thing out of the way now!

phated commented:
Awesome, thanks!

phated commented:
:shipit: This will be 2.0 due to the coercer behavior change

phated commented:
Published.  Great work as always!

erikkemperman commented:
Thanks for review and merge!
